src/platform/datapath_raw_xdp.c

```diff
@@ -788,6 +803,93 @@ CxPlatDpRawInterfaceUpdateRules(
     }
 }
 
+_IRQL_requires_max_(PASSIVE_LEVEL)
+void
+CxPlatDpRawInterfaceAddRule(
+    _In_ XDP_INTERFACE* Interface,
+    _In_ const XDP_RULE* NewRule
+    )
+{
+#pragma warning(push)
+#pragma warning(disable:6386) // Buffer overrun while writing to 'NewRules' - FALSE POSITIVE
+
+    CxPlatLockAcquire(&Interface->RuleLock);
+    // TODO - Don't always allocate a new array?
+
+    if (Interface->RuleCount + 1 == 0) {
+        QuicTraceEvent(
+            LibraryError,
+            "[ lib] ERROR, %s.",
+            "No more room for rules");
+        CxPlatLockRelease(&Interface->RuleLock);
+        return;
+    }
+
+    const size_t OldSize = sizeof(XDP_RULE) * (size_t)Interface->RuleCount;
+    const size_t NewSize = sizeof(XDP_RULE) * ((size_t)Interface->RuleCount + 1);
+
+    XDP_RULE* NewRules = CxPlatAlloc(NewSize, RULE_TAG);
+    if (NewRules == NULL) {
+        QuicTraceEvent(
+            AllocFailure,
+            "Allocation of '%s' failed. (%llu bytes)",
+            "XDP_RULE",
+            NewSize);
+        CxPlatLockRelease(&Interface->RuleLock);
+        return;
+    }
+
+    if (Interface->RuleCount > 0) {
+        memcpy(NewRules, Interface->Rules, OldSize);
+    }
+    NewRules[Interface->RuleCount] = *NewRule;
+    Interface->RuleCount++;
+
+    if (Interface->Rules != NULL) {
+        CxPlatFree(Interface->Rules, RULE_TAG);
+    }
+    Interface->Rules = NewRules;
+
+    CxPlatDpRawInterfaceUpdateRules(Interface);
+
+    CxPlatLockRelease(&Interface->RuleLock);
+
+#pragma warning(pop)
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+void
+CxPlatDpRawInterfaceRemoveRule(
+    _In_ XDP_INTERFACE* Interface,
+    _In_ const XDP_RULE* Rule
+    )
+{
+    CxPlatLockAcquire(&Interface->RuleLock);
+
+    for (uint8_t i = 0; i < Interface->RuleCount; i++) {
+        if (Interface->Rules[i].Match != Rule->Match) {
+            continue;
+        }
+
+        if (Rule->Match == XDP_MATCH_UDP_DST) {
+            if (Rule->Pattern.Port != Interface->Rules[i].Pattern.Port) {
+                continue;
+            }
```

Comment:
This is probably not for long term but I believe we can bring down the time complexity of add/remove to O(1) or amortized O(1).

Reply:
I'm not worried too much about perf here right now.